New integer protocols

dabrahams opened the pull request:
Work on integers from @moiseev 
#### What's in this pull request?

<!-- Description about pull request. -->
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

A smoke test on macOS does the following:
1. Builds the compiler incrementally.
2. Builds the standard library only for macOS. Simulator standard libraries and
   device standard libraries are not built.
3. lldb is not built.
4. The test and validation-test targets are run only for macOS. The optimized
   version of these tests are not run.

A smoke test on Linux does the following:
1. Builds the compiler incrementally.
2. Builds the standard library incrementally.
3. lldb is built incrementally.
4. The swift test and validation-test targets are run. The optimized version of these
   tests are not run.
5. lldb is tested.

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

Commit:
no IntMax in IntegerParsing

Commit:
Addressing some of the FIXMEs

dabrahams commented on stdlib/private/SwiftPrivate/SwiftPrivate.swift:
```diff
@@ -13,14 +13,15 @@
 /// Convert the given numeric value to a hexadecimal string.
-public func asHex<T : Integer>(_ x: T) -> String {
-  return "0x" + String(x.toIntMax(), radix: 16)
+public func asHex<T : FixedWidthInteger>(_ x: T) -> String {
```
Can’t we convert an arbitrary-width integer to hex?

moiseev replied:
Added a `FIXME`.

dabrahams commented on stdlib/public/SDK/SceneKit/SceneKit.swift:
```diff
@@ -151,7 +151,7 @@ extension SCNGeometryElement {
   /// that has a constant number of indices per primitive
   /// - Precondition: the `primitiveType` must be `.triangles`, `.triangleStrip`, `.line` or `.point`
-  public convenience init<IndexType : Integer>(
+  public convenience init<IndexType : FixedWidthInteger>(
```
Why does this require fixed width?

moiseev replied:
It does not require anything. Can as well be `BinaryInteger`, but I think the underlying implementation will be quite surprised if given something with a random memory representation, i.e. a 3rd party bignum.

dabrahams replied:
Oh!  Heck, their constraint is wrong to begin with.  They are assuming all `Integer`s are POD.  We should file a radar.

dabrahams commented on stdlib/public/core/FloatingPoint.swift.gyb:
```diff
@@ -834,12 +829,16 @@ extension FloatingPoint {
   }
 %end
 
+// FIXME(integers): maybe uncomment it back
```
Why bring this back?  If there’s a possible reason, the FIXME should note it

dabrahams commented on stdlib/public/core/IntegerParsing.swift.gyb:
```diff
@@ -30,9 +27,9 @@ UIntMax = 'UInt%s' % int_max_bits
-internal func _parseUnsignedAsciiAsUIntMax(
-  _ u16: String.UTF16View, _ radix: Int, _ maximum: UIntMax
-) -> UIntMax? {
+internal func _parseUnsignedAsciiAsUInt64(
```
Seems to me that we want these all to be generic over every BinaryInteger type, neh?

moiseev replied:
Yes. But changes like this a better done when the whole thing at least compiles.

dabrahams replied:
Then you need to leave a FIXME in place, right?

moiseev replied:
Added a few [here](https://github.com/apple/swift/commit/3c0e9d534fabd2ce3b90f050b36873745c205d08).

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -14,6 +14,8 @@
 # Utility code for later in this template
 #
 
+from SwiftIntTypes import all_integer_types, int_max_bits, should_define_truncating_bit_pattern_init
```
Are we keeping the `truncating:` init?  Since we have `truncatingOrExtending:` it seems slightly unnecessary.  Though it could help expressivity...

moiseev replied:
There is a FIXME elsewhere to consider merging them.

dabrahams commented:
Lots of missing doc comments in the code!

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -237,11 +255,14 @@ extension BinaryInteger {
   public func signum() -> Self {
-    // FIXME(integers): implement
-    fatalError()
+    if self < (0 as Self) { return -1 }
```
if these `as Self`s are really needed I think there’s a compiler bug that should be reported.  Otherwise we should just say `0`.

moiseev replied:
They were needed at some point, but are not now. Removed.

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -260,11 +306,22 @@ public func == <T : BinaryInteger>(lhs: T, rhs: T) -> Bool {
   return lhs.isLess(than: rhs)
 }
 
+// FIXME(integers): these seem to be important for Int-backed enums
```
important in what way?

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -283,8 +340,8 @@ public func != <T : BinaryInteger, U : BinaryInteger>(lhs:T, rhs: U) -> Bool {
-  let lhsSign = lhs < 0 ? -1 : lhs > 0 ? 1 : 0
-  let rhsSign = rhs < 0 ? -1 : rhs > 0 ? 1 : 0
+  let lhsSign = lhs < (0 as T) ? -1 : lhs > (0 as T) ? 1 : 0
+  let rhsSign = rhs < (0 as U) ? -1 : rhs > (0 as U) ? 1 : 0
```
If these are ambiguous without the casts, we should make it a requirement of BinaryInteger that there be a (Self,Self) version of the comparison.

moiseev replied:
These are not ambiguous, it's the inlining. It is trying to inline itself. Type cast makes it call the non-generic version, and avoid the problem.
We already have a (Self, Self) comparison requirement on the `BinaryInteger`, it's just called `isLess(than:)`. Given [SE-0091](https://github.com/apple/swift-evolution/blob/master/proposals/0091-improving-operators-in-protocols.md) it will become a `static func < (Self, Self) -> Bool`.
For now, changing explicit type cast to using `isLess(than:)`.

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -423,17 +480,17 @@ public func ${x.operator}= <T: FixedWidthInteger>(lhs: inout T, rhs: T) {
 % for x in maskingShifts:
 
-@_transparent
+//@_transparent
```
Why are these commented out?

moiseev replied:
There was a inlining loop that I decided to avoid first and fix later.

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -515,22 +568,21 @@ extension FixedWidthInteger {
   @_transparent
-  public func unsafe${capitalize(x.name)}(rhs: Self) -> Self {
-    let (result, overflow) = self.${x.name}WithOverflow(${callLabel}rhs)
+  public func unsafe${capitalize(x.name)}(other: Self) -> Self {
```
This should be a label-less argument, shouldn’t it?

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -447,41 +504,37 @@ public func ${x.operator}= <
-  //if _slowPath(rhs <= -T.bitWidth) {
-    //return overshift${'RL'['R' in x.name]}
-  //}
-  //return lhs ${x.operator.translate(maketrans('<>', '><'))} -rhs
```
It’s not exactly clear to me what the basis for this comparison is, or how to evaluate the changes.  Could you clarify?

moiseev replied:
The implementation is copied verbatim from the integers prototype, the last known state of which is available [here](https://github.com/apple/swift/blob/b7622b41756e33bdf3f3415320ffec52aec73281/test/Prototypes/Integers.swift.gyb).

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -605,59 +648,66 @@ public func &${x.operator} <T: FixedWidthInteger>(lhs: T, rhs: T) -> T {
+      x /= 10
+      buf.append(
+        UnicodeScalar(
+          ascii0 + Int(UInt(extendingOrTruncating: r)._value))!)
```
This looks too complicated; there ought to be a simpler way to express whatever the intention is, I think.

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -605,59 +648,66 @@ public func &${x.operator} <T: FixedWidthInteger>(lhs: T, rhs: T) -> T {
     // FIXME(integers): uncomment checks
-    //_assertCond(
-      //source >= 0, "negative value \(source) not representable by \(Self.self)")
+    //_precondition(source >= 0,
```
The old checks provide better user experience and I *think* _assertCond might be something the performance team can optimize better.  We should check with them before making the change to _precondition.

moiseev replied:
Optimizer team explicitly advised against `_assertCond`. We can it leave it to them to replace with something more appropriate though...

dabrahams commented on stdlib/public/core/Integers.swift.gyb:
```diff
@@ -727,3 +776,440 @@ extension SignedInteger_ where Self : FixedWidthInteger {
+  return U(x)
+}
+
+internal func _unsafePlus(_ lhs: Int, _ rhs: Int) -> Int {
```
what are these for?

moiseev replied:
They are used elsewhere, and I just copied them from the old implementation to make the whole thing compile. The plan is to replace them with the method call, since we have the unsafeOP methods on `FixedWidthInteger`.

dabrahams commented on stdlib/public/core/Slice.swift.gyb:
```diff
@@ -395,7 +395,7 @@ public struct ${Self}<Base : ${BaseRequirements}>
       let shouldUpdateStartIndex = bounds.lowerBound == _startIndex
       let lastValidIndex = _base.index(before: bounds.lowerBound)
-      let newEndIndexOffset =
+      let newEndIndexOffset: Base.IndexDistance =
```
Why isn’t this type deduced?

dabrahams commented on stdlib/public/core/Sort.swift.gyb:
```diff
@@ -140,7 +140,7 @@ func _introSort<C>(
 %   end
   let count =
-    elements.distance(from: range.lowerBound, to: range.upperBound).toIntMax()
+    elements.distance(from: range.lowerBound, to: range.upperBound)
```
Maybe just drop the named constant now that we can do the comparison without type coercion.

dabrahams commented on stdlib/public/core/StaticString.swift:
```diff
@@ -160,7 +160,9 @@ public struct StaticString
-    self._flags = Bool(isASCII) ? (0x2 as UInt8)._value : (0x0 as UInt8)._value
+    self._flags = Bool(isASCII)
+      ? (0x2 as UInt8)._value
+      : (0x0 as UInt8)._value
```
Can we do

```swift
self._flags = isASCII ? 0x2._value : 0x0._value
```

?

dabrahams commented on stdlib/public/core/StringLegacy.swift:
```diff
@@ -238,8 +238,8 @@ extension String {
-  public init<T : _SignedInteger>(_ v: T) {
-    self = _int64ToString(v.toIntMax())
+  public init<T : FixedWidthInteger>(_ v: T) {
+    self = _int64ToString(Int64(v))
```
This looks wrong; the constraint should be more general, and it should work if the value is wider than 64 bits.

moiseev replied:
Yes. But should it be part of this change? 64 bits is maximum we have so far, so I figured it was sufficient this way. At least for now, given time constraints.

dabrahams commented on stdlib/public/core/StringLegacy.swift:
```diff
@@ -275,12 +276,12 @@ extension String {
     _precondition(radix > 1, "Radix must be greater than 1")
     self = _int64ToString(
-      value.toIntMax(), radix: Int64(radix), uppercase: uppercase)
+      Int64(value), radix: Int64(radix), uppercase: uppercase)
```
Also looks wrong for the same reasons.

dabrahams commented on stdlib/public/core/Unicode.swift:
```diff
@@ -344,7 +344,7 @@ public struct UTF8 : UnicodeCodec {
       }
       // Extract data bits.
-      // FIXME(integers): remove extra type casts
+      // FIXME(integers): remove extra type hints
```
I think we can use masking shifts here, too?

dabrahams commented on stdlib/public/core/UnsafeRawPointer.swift.gyb:
```diff
@@ -410,7 +410,7 @@ public struct Unsafe${Mutable}RawPointer : Strideable, Hashable, _Pointer {
       Builtin.int_memcpy_RawPointer_RawPointer_Int64(
         (self + offset)._rawValue, rawSrc, UInt64(sizeof(T.self))._value,
-        /*alignment:*/ Int32(0)._value,
+        /*alignment:*/ Int32()._value,
```
Does `0._value` work?

Commit:
adding a proper arguemnt label to 'unsafe' methods

Commit:
replacing Int32()._value with let zero: Int32 = 0

moiseev commented on stdlib/public/core/CTypes.swift:
```diff
@@ -193,9 +193,10 @@ func _memcpy(
   Builtin.int_memcpy_RawPointer_RawPointer_Int64(
     dest, src, size,
-    /*alignment:*/ Int32()._value,
+    /*alignment:*/ zero._value,
```
Using `0._value` makes it an `Int64`. Changing the occurrences to `let zero: Int32 = 0`.

dabrahams replied:
Oh, type constraints on the parameters to `Builtin.int_memcpy_RawPointer_RawPointer_Int64` don’t handle the deduction?  That’s weird.  It’s not like any `Builtin` is overloaded.

Commit:
removing commented out default implementation

Commit:
FIXMEs to change FixedWidthInteger to BinaryInteger in generic code

Commit:
removing explicit typecasts

Commit:
FIXME for the unsafe functions

Commit:
a better fixme message on == (Int, Int). removing < (Int, Int)

Commit:
removing IntegerArithmetic.swift

Commit:
implementing popcount

Commit:
removing unnecessary fatalError calls

Commit:
a bunch of unavailable declarations

Commit:
fixing the benchmarks

Commit:
Turning operators into static funcs

Commit:
Simplifying too complex expressions in benchmarks with type casts

Commit:
getting rid of toIntMax in StdlibCollectionUnittest

Commit:
Merge 'origin/master' into new-integer-protocols

Commit:
[stdlib] Correct renamed protocol @available decls

Commit:
[emacs support] properly parse some test failures

Commit:
[stdlib] Improve unavailable diagnostics

Making these generic and removing inout from the arguments cuts down on
the number of useless notes we give users when they try to use them.
Arguably the compiler should do better, but this moves our test results
along.

Commit:
[stdlib] Fix constraint tests for new integers

Commit:
[stdlib] New integers simplify generic code!

Commit:
Provisionally adjust DebugInfo test

karwa commented:
Just to check: this is definitely going to make Swift 3.0, right?

gribozavr commented:
> Just to check: this is definitely going to make Swift 3.0, right?

Right now it looks like it will not.  There are many test failures that haven't been solved, we noticed performance issues, and the implementation work is not even complete.

natecook1000 committed:
[Integer protocols] Make .negate() the customization point (#4413)

The SignedArithmetic protocol provides default implementations for both
negate() and negated(), with negate() calling negated(). This is inconsistent
with the Arithmetic protocol, which provides implementations of the nonmutating
methods that call their mutating counterparts. This flips the implementations
of negate() and negated() so an implementer only has to focus on mutating
methods.

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
[integers] making stdlib compile after merging master

Commit:
[integers] adding labels to the quotientAndRemainder return tuple

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
turning the memcpy and memmove alignment arguments into compile time constants

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Adding a (bad) implementation of init(_ source: FloatingPoint) to fix some tests

Commit:
introducing concrete initializers from all floating point types

Commit:
Merge branch 'master' into new-integer-protocols

natecook1000 committed:
[stdlib] Revise documentation for integer types & protocols (#5083)

* [stdlib] Revise documentation for integer types & protocols

* Update with feedback

* [stdlib] Additional integer documentation revisions

Commit:
[WIP] fixing some of the integer related test failures

Commit:
Merge branch 'new-integer-protocols' of github.com:apple/swift into new-integer-protocols

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Operators as static funcs for Arithmetic (see SE-0091)

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Moving unavailable operators out of concrete types

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Adding specific versions of arithmetics operators for CGFloat to speed up compilation

Commit:
Revert "Adding specific versions of arithmetics operators for CGFloat to speed up compilation"

This reverts commit c91dce7394404e26e1470f7217726522ccca8356.

Commit:
Breaking up complex expression to improve compilation time

Commit:
Cleaning up the gyb

Commit:
Removing FIXMEs

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Comparable operators to static funcs

Commit:
Choosing the right overloads for comparison operators

Commit:
Minor fixes and improvements

Commit:
Avoiding the force unwrap in UnsafePointer.count

Commit:
Turning more operators into static funcs

Commit:
Get rid of _DisallowMixedSignArithmetic

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Adding back accidentally deleted UnsignedInteger protocol

Commit:
Using Strideable APIs instead of + and -

Commit:
Different Strideable conformances for Swift 3 and Swift 4

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Fixing Dispatch tests

Commit:
Operators from free functions to static methods

Commit:
Obsoleted concrete type overloads for integer arithemtic operators

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
@swift-ci Please benchmark

Commit:
Adding a few missing operators

Commit:
Commenting out the arithmetic operators on concrete types

Commit:
Adding a temporary workaround for a missing optimizer feature

Commit:
Commenting out some checks pending optimizer work

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

moiseev commented:
@swift-ci Please benchmark

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
@swift-ci Please benchmark

Commit:
Marking workarounds for <rdar://problem/29029561>

Commit:
Addressing compiler warning for unsigned integers

Commit:
Marking init(truncatingBitPattern:) as obsoleted in Swift 4

Commit:
Marking IntMax and UIntMax as obsoleted

Commit:
numericCast is wort keeping

Commit:
Turning SignedArithmetic.negated() into a static prefix func -

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Removing workarounds for <rdar://problem/29029561>

moiseev commented:
@swift-ci Please benchmark

Commit:
SeeAlso updated for the negate() function

Commit:
Uncommenting arithmetic operators in the FloatingPoint protocol

Commit:
Proper availability attribute for toIntMax() method

Commit:
Implementing doubleWidthDivide

Commit:
Speeding up Strideable conformance for integers

Commit:
Changing preconditions in integer inits for better performance

Commit:
Adding a fixme to possibly remove minimumSignedRepresentationBitWidth

Commit:
Revert "Speeding up Strideable conformance for integers"

This reverts commit 8789e627f62f28d6045c1a427294a268dc7d6706.

Commit:
Specialized implementation of Strideable conformance for Int

Commit:
[SILOptimizer] add support for always false comparisons to ArrayBoundsCheckOpts

dabrahams reviewed (commented):
##

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
using masking shifts

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Using more masking shifts

Commit:
Hiding the shift operators of type (Self, Int) -> Self

Commit:
Strideable for _Pointer is not necessary in Swift 4

All the necessary methods and operators are already implemented for
concrete pointer types more efficiently.

Commit:
Simplifying masking shift implementation on concrete integers

Commit:
Using faster initializer for integers where appropriate

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Speeding up the RC4 benchmark by using right initializers for ints

Commit:
Revert "Speeding up the RC4 benchmark by using right initializers for ints"

This reverts commit 4f396089d95f1e0abdc370c1eda60694b4793c38.

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
optimizing preconditions

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Faster comparisons with integer literals

Commit:
Avoiding recursive calls to init(extendingOrTruncating:) and speeding up ==

Commit:
Minor improvements

moiseev commented:
@swift-ci Please benchmark

Commit:
adding type hints in signum implementation

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Removing unary + for floating point types and changing the unary - to a static method

Addresses: <rdar://problem/16424438>

Commit:
Fixing the benchmark crash

... by removing the + - overloads for Strideable, and only leaving them
for pointers (both Strideable and _Pointer).

Strideable was used in the past to simplify advancing collection
indices, but with new collection indexing model it is no longer
necessary to overload `+` and `-` for all strideable types..

Commit:
Changing == and < on UnsafeRawPointer to static funcs

Commit:
Fixing the benchmarks now that the + is not defined for Strideable

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Making StdlibUnittest compile in absense of + on Strideable

Commit:
Not using arithmetic operators on strideables in test

Commit:
Deprecating the BitwiseOperations protocol in favor of FixedWidthInteger

Commit:
Adding trailingZeros property to BinaryInteger protocol

<rdar://problem/17505796>

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Deprecation message to FixedWidthInteger.allZeros

Commit:
Consistent implementation of mutating vs non-mutating operators on FixedWidthInteger

Commit:
Adding prefix ~ protocol requirement to FixedWidthInteger

Commit:
Type hints for operators

Commit:
More type hints and inlining to speed up integer convertions

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Speeding up conversion between integer types

This change levereges the fact that there is no need to check for the
upper bound if the source type if narrower than the destination type.

Commit:
Speeding up .description for built-in integers

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Speeding up the heterogeneous == and < by using more opimizable checks

Commit:
Cleaning up the smart shifts

Commit:
Commenting an overload that leads to compiler error

Commit:
Removing overloads of == on concrete types in favor of a generic one that is now faster

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Fixing a typo

Commit:
Speeding up the generic smart shift

Commit:
Revert "Speeding up the generic smart shift"

This reverts commit 26111e8dd8d6a180de7912926529878104d0058a.

Commit:
Adding integer unit tests from the original prototype

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Adding some doubleWidthMultiply and bouldWidthDivide tests

Commit:
Using init(extendingOrTruncating:) instead of a deprecated init(truncatingBitPattern:)

Commit:
Using masking shifts

Commit:
Removing dead and commented code

Commit:
Type hints in comparisons

Commit:
Overflow checks in division/modulo operators + tests

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
doubleWidth operations on Int64 need special implementation for 32bit platforms

Commit:
Moving bitwise operations and shifts to BinaryInteger protocol

Commit:
Implementing non-mutating shifts using mutating ones (for consistency)

Commit:
Adding smart shift default implementations to BinaryInteger via an extension

Commit:
Getting rid of no longer needed minimumSignedRepresentationBitWidth

Commit:
Adding endianness related inits and properties to FixedWidthInteger

Commit:
Introducing abs<T : SignedArithmetic & Comparable>

Commit:
Adding DoubleWidth<>

Commit:
Correcting doubleWidthDivide precondition

Commit:
Adding init(bitPattern: FLOATING_POINT) to Int32 and Int64

Commit:
Using trailingZeros and leadingZeros provided by integers in floating point types

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Fixing some and XFAILing other failing tests

Commit:
Fixing stdlib tests

Commit:
More test fixes

Commit:
Disabling some SILGen tests, that started to fail due to new integers

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Fixing and XFAILing tests

Commit:
Fixing couple more tests

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Making non-masking shift helper function public (compiler requires it now)

Commit:
Removing explicit implementation of arithmetic operators from Decimal in favor of defaults

Commit:
Fixing more tests

Commit:
Getting rid of Arithmetic.init() in favor of 0

Commit:
Annotating init() as transparent

Commit:
Using chain() instead of a 2-level list comprehension in gyb

Commit:
Moving arithmetic operators back onto concrete types level temporarily

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Removing arithemtic operator default implementations from BinaryInteger and FloatingPoint

Commit:
Implementing arithmetic operators on NSDecimal, CGFloat, and concrete FP types

Commit:
Fixing tests

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Fixing tests with artihmetic operators back on concrete types

Commit:
Using >= instead of > to catch overshifts correctly

Commit:
Correcting the logic in the reasoning for the heterogeneous comparison implementation

natecook1000 committed:
Add BigInt prototype for testing generic integers (#7005)

* [new-integer-protocols] Add BigInt prototype

* Update to use 2s complement for bitwise operations

* Add Bit type and tests using BigInt<Bit>

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Adding back a transparrent attribute in an initializer

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Fixing the tests (again)

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Moving / down to the BinaryInteger protocol

Commit:
Renaming leadingZeros to leadingZeroBits. Same for trailingZeros

Commit:
Fixing the trailingZeroBits test

natecook1000 committed:
[stdlib] Doc comment revisions for (un)signed integers (#7541)

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Merge remote-tracking branch 'origin/new-integer-protocols' into new-integer-protocols

Commit:
Fixing heterogeneous <, failable init from other ints, and init from floats

Commit:
Fixing some validation tests

Commit:
Arithmetic => Numeric

Commit:
Properly naming bit count properties

moiseev commented:
@swift-ci Please test macOS platform

Commit:
WithOverflow => ReportingOverflow

moiseev commented:
Infrastructure failure. Let's try it again.
@swift-ci Please test macOS platform

shahmishal commented:
@swift-ci Please test macOS

shahmishal commented:
@swift-ci Please smoke test macOS

Commit:
multipliedFullWidth(by:) and dividingFullWidth(_:)

Commit:
Default implementation of instance bitWidth for FixedWidthInteger

Conforming types already provide the static bitWidth property. Instance
one can be implemented in a protocol extension.

Resolves: <rdar://problem/30186638>

Commit:
Fixing some more validation tests

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

Commit:
Removing the test artifacts

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

moiseev commented:
@swift-ci Please test macOS platform

shahmishal commented:
Updating Xcode on CI, I will re-trigger testing after Xcode has been installed.

Commit:
Fixing the rest of the tests

Commit:
Fixing tests for Linux

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
@swift-ci Please test macOS platform

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Improving benchmarks

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Improving the performance of String.UTF8View

Commit:
Adding an explicit type context for ==

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
@swift-ci Please test macOS platform

Commit:
Fixing a few SIL tests

Commit:
Merge branch 'master' into new-integer-protocols

shahmishal commented:
@swift-ci Please clean test macOS platform

shahmishal commented:
```
PlaygroundLogger/PlaygroundLogger/Common.swift:70:11: error: use of undeclared type 'Integer'
```

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
https://github.com/apple/swift-xcode-playground-support/pull/13
@swift-ci Please test macOS platform

moiseev commented:
https://github.com/apple/swift-xcode-playground-support/pull/13
@swift-ci Please test macOS platform

moiseev commented:
https://github.com/apple/swift-xcode-playground-support/pull/13
@swift-ci Please clean test macOS platform

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please clean test macOS platform

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please clean test macOS platform

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please clean test macOS platform

Commit:
Speeding up the build by breaking complex expressions into smaller pieces

Commit:
Adding an explicit type context

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Reverting becnhmarks code to master

Commit:
Replacing @_transparent with @inline(__always)

Mandatory inlining of a few very frequently used functions caused the
SIL size explosion, which in turn made the inliner use up to 6Gb of
memory to compile the standard library. @inline(__always) helps avoid
that without affecting benchmark results.

Related to: <rdar://problem/31375011>

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

shahmishal commented:
```
01:45:36.729 Failing Tests (2):
01:45:36.729     Swift(macosx-x86_64) :: api-digester/source-stability.swift
01:45:36.729     Swift(macosx-x86_64) :: DebugInfo/returnlocation.swift
```

moiseev commented:
It did not blow up the CI bot! 🍾🎉

moiseev commented:
@swift-ci Please benchmark

Commit:
Fixing the returnlocation test

Commit:
Merge branch 'master' into new-integer-protocols

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

moiseev commented:
@swift-ci Please benchmark

Commit:
Adding a few Swift 3 compatibilty APIs

Commit:
Adding deprecated mixed-type arithmetic operators for Strideable & SignedInteger for compatibility

Commit:
Arithmetic operators for Strideable where Stride : ExpressibleByIntegerLiteral for compatibility

Commit:
Adding overloads for << and >> with the literal on the left-hand side

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Splitting a complex expression in the benchmarks

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

moiseev commented:
@swift-ci Please benchmark

Commit:
Speculatively replacing a couple more transparent's with inline-always'es

Commit:
Adding non-masking shifts on concrete types

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Adding SignedInteger.&+ and &- for source compatibility

Commit:
More typealiases for Swift 3 protocols

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Removing obsolete rename tests

Commit:
Commenting out deprecated +/- on Strideable

Commit:
Running new integer tests in Swift 4 mode only

Commit:
Handling the differences in << and >> for Swift 3 vs Swift 4

In Swift 3 shifts used to be defined on the concrete integer types, so
the right-hand-side value in the shift expression could define a type
for the result, as in `1 << i32` would have the type Int32. Swift 4
makes shift operators heterogeneous, so now `1 << i32` will result in an
Int, according to the type of the left-hand-side value, which gets a
default type for integer literals.

Commit:
Overloads for +/- on Strideable for Swift 3 compatibility mode

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

moiseev commented:
@swift-ci Please benchmark

Commit:
Splitting the MixedTypeArithmeticsDiagnostics test into two (Swift 3 and Swift 4)

Commit:
Fixing some tests

Commit:
Fixing the unicode tests by removing the potential optimization

Commit:
Listing expected source stability exceptions

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

moiseev commented:
https://github.com/apple/swift-corelibs-foundation/pull/951
https://github.com/apple/swift-corelibs-xctest/pull/187
@swift-ci Please test Linux platform

Commit:
Merge branch 'master' into new-integer-protocols

Commit:
Fixing the complex expression test

Commit:
Adding arithmetic methods to FloatingPoint for Swift 3 mode

Commit:
Adding Swift 3 alias for _SignedInteger

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

Commit:
Fixing the iOS test failures

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

Commit:
Merge remote-tracking branch 'origin/master' into new-integer-protocols

moiseev commented:
apple/swift-xcode-playground-support#13
@swift-ci Please test macOS platform

moiseev commented:
apple/swift-corelibs-foundation#951
apple/swift-corelibs-xctest#187
@swift-ci Please test Linux platform

DougGregor commented:
W00t!

dabrahams commented:
@moiseev https://github.com/dabrahams/swift/tree/integer-merge is the beginning of a merge into the unicode-rethink branch.  Everything seems good except for AlternateIntegerWidths.swift.gyb, which you can experiment with in isolation if you like: https://gist.github.com/dabrahams/f7fa2b710b50340e168926842515de4b
